src/protocol/aggregation/mod.rs

```diff
@@ -52,139 +58,229 @@ where
     BK: GaloisField,
 {
     let validator = sh_ctx.narrow(&Step::Validator).validator::<F>();
+    let ctx = validator.context();
 
-    let (gf2_value_bits, _gf2_breakdown_keys) = (
-        get_gf2_value_bits(input_rows),
-        get_gf2_breakdown_key_bits(input_rows),
+    // convert the input from `[Z2]^u` into `[Zp]^u`
+    let (converted_value_bits, converted_breakdown_key_bits) = (
+        upgrade_bit_shares(
+            ctx.narrow(&Step::ConvertValueBits),
+            input_rows,
+            CV::BITS,
+            |row, i| {
+                Replicated::new(
+                    Gf2::truncate_from(row.contribution_value.left()[i]),
+                    Gf2::truncate_from(row.contribution_value.right()[i]),
+                )
+            },
+        )
+        .await?,
+        upgrade_bit_shares(
```

Comment:
This can be done in parallel with the line above.

Reply:
Changed to use `try_join` on the tuples.